ring/batch.go

```diff
@@ -112,29 +129,45 @@ func DoBatch(ctx context.Context, op Operation, r ReadRing, keys []uint32, callb
 }
 
 func (b *batchTracker) record(sampleTrackers []*itemTracker, err error) {
-	// If we succeed, decrement each sample's pending count by one.  If we reach
-	// the required number of successful puts on this sample, then decrement the
-	// number of pending samples by one.  If we successfully push all samples to
-	// min success instances, wake up the waiting rpc so it can return early.
-	// Similarly, track the number of errors, and if it exceeds maxFailures
-	// shortcut the waiting rpc.
+	// If we reach the required number of successful puts on this sample, then decrement the
+	// number of pending samples by one.
 	//
-	// The use of atomic increments here guarantees only a single sendSamples
-	// goroutine will write to either channel.
+	// The use of atomic increments here is needed as:
+	// * rpcsPending and rpcsFailed guarantees only a single sendSamples goroutine will write to either channel
+	// * succeeded, failed4xx, failed5xx and remaining guarantees that the "return decision" is made atomically
+	// avoiding race condition
 	for i := range sampleTrackers {
 		if err != nil {
-			if sampleTrackers[i].failed.Inc() <= int32(sampleTrackers[i].maxFailures) {
-				continue
-			}
-			if b.rpcsFailed.Inc() == 1 {
-				b.err <- err
+			// Track the number of errors by error family, and if it exceeds maxFailures
+			// shortcut the waiting rpc.
+			errCount := sampleTrackers[i].recordError(err)
+			// We should return an error if we reach the maxFailure (quorum) on a given error family OR
+			// we dont have any remaining instances to try
```

Comment:
It would help (me) to have a comment which explains why we want to count the 4xx and 5xx separately

Reply:
I tried explaining it better in f2e37f516dee91b324371a49c452ed34df723269. Let me know if that makes it more clear.

Reply:
thanks, that helps!